### ISSUE TITLE
BasicIntellisense fails in ETA.

### ISSUE THREAD

**Issue by heejaechang** [dotnet/roslyn#9551]
PR: [#648](https://github.com/dotnet/roslyn-internal/pull/648) *Merge master into future* by @vslsnap
Failure: http://dotnet-ci.cloudapp.net/job/roslyn-internal_prtest_win_dbg_eta/741/

**3 Test Failures:**
BasicIntellisense.Signature help shows up
SplitFile.VB keyword highlighting
SplitFile.C# Formatting

**Issue 1: Individual TAO Test Failure**
Failing integration tests: 

BasicIntellisense.xml

SplitFile.xml

**Comment by davkean** [comment]
Split file is dupe of https://github.com/dotnet/roslyn/issues/9454

**Comment by DustinCampbell** [comment]
@balajikris, could you take a look at BasicIntelliSense?

**Comment by balajikris** [comment]
This is another instance where ETA caught and logged an exception in the background and moved on..
From the dump, **ETA** is trying to update progression UI from the background thread and WPF duly throws.  What Im not sure yet is why this would only occasionally fail. This isn't a product bug btw.

~~~
0:000> !pe 11b47fe0
Exception object: 11b47fe0
Exception type:   System.InvalidOperationException
Message:          The calling thread cannot access this object because a different thread owns it.
InnerException:   <none>
StackTrace (generated):
    SP       IP       Function
    16CCE1F0 07B7F46F WindowsBase!System.Windows.Threading.Dispatcher.VerifyAccess()+0x47
    16CCE1FC 1456F8E0 PresentationCore!System.Windows.Media.Imaging.BitmapSource+WeakBitmapSourceEventSink.AttachSourceDownloadHandlers()+0x18
    16CCE20C 1456F8B0 PresentationCore!System.Windows.Media.Imaging.BitmapSource+WeakBitmapSourceEventSink.set_EventSource(System.Windows.Media.Imaging.BitmapSource)+0x30
    16CCE21C 1456F864 PresentationCore!System.Windows.Media.Imaging.BitmapSource.RegisterDownloadEventSource(System.Windows.Media.Imaging.BitmapSource)+0x3c
    16CCE230 1456F7FC PresentationCore!System.Windows.Media.Imaging.CroppedBitmap.SourcePropertyChangedHook(System.Windows.DependencyPropertyChangedEventArgs)+0x3c
    16CCE25C 1456F77B PresentationCore!System.Windows.Media.Imaging.CroppedBitmap.SourcePropertyChanged(System.Windows.DependencyObject, System.Windows.DependencyPropertyChangedEventArgs)+0x43
    16CCE284 0804BB8D WindowsBase!System.Windows.DependencyObject.OnPropertyChanged(System.Windows.DependencyPropertyChangedEventArgs)+0x4d
    16CCE2B4 0804BB0D WindowsBase!System.Windows.Freezable.OnPropertyChanged(System.Windows.DependencyPropertyChangedEventArgs)+0x25
    16CCE2DC 0804B8DE WindowsBase!System.Windows.DependencyObject.NotifyPropertyChange(System.Windows.DependencyPropertyChangedEventArgs)+0x3e
    16CCE330 08050BA5 WindowsBase!System.Windows.DependencyObject.UpdateEffectiveValue(System.Windows.EntryIndex, System.Windows.DependencyProperty, System.Windows.PropertyMetadata, System.Windows.EffectiveValueEntry, System.Windows.EffectiveValueEntry ByRef, Boolean, Boolean, System.Windows.OperationType)+0x7b5
    16CCE3E4 07B7FAF9 WindowsBase!System.Windows.DependencyObject.SetValueCommon(System.Windows.DependencyProperty, System.Object, System.Windows.PropertyMetadata, Boolean, Boolean, System.Windows.OperationType, Boolean)+0x4e1
    16CCE44C 07B7F410 WindowsBase!System.Windows.DependencyObject.SetValueInternal(System.Windows.DependencyProperty, System.Object)+0x38
    16CCE460 1456F1FC PresentationCore!System.Windows.Media.Imaging.CroppedBitmap..ctor(System.Windows.Media.Imaging.BitmapSource, System.Windows.Int32Rect)+0x44
    16CCE48C 1E4ED58E Microsoft_VisualStudio_Platform_VSEditor!Microsoft.VisualStudio.Language.Intellisense.Implementation.StandardGlyphProvider.GetGlyph(Microsoft.VisualStudio.Language.Intellisense.StandardGlyphGroup, Microsoft.VisualStudio.Language.Intellisense.StandardGlyphItem)+0xde
    16CCE4C0 1456F71E Microsoft_VisualStudio_LanguageServices!Microsoft.VisualStudio.LanguageServices.Implementation.Progression.IconHelper+<>c__DisplayClass2_1.<Initialize>b__0()+0x46
    16CCE4D8 1456F6A8 EditorTestApp!Roslyn.Editor.TestApp.UI.ToolWindows.GraphNodeItemHost.Microsoft.VisualStudio.Progression.IIconService.AddIcon(System.String, System.String, System.Func`1<System.Windows.Media.ImageSource>)+0x40
    16CCE4FC 1456B672 Microsoft_VisualStudio_LanguageServices!Microsoft.VisualStudio.LanguageServices.Implementation.Progression.IconHelper.Initialize(Microsoft.VisualStudio.Language.Intellisense.IGlyphService, Microsoft.VisualStudio.Progression.IIconService)+0x50a
    16CCE5F4 14569E2F Microsoft_VisualStudio_LanguageServices!Microsoft.VisualStudio.LanguageServices.Implementation.Progression.AbstractGraphProvider.EnsureInitialized()+0x97
    16CCE618 14569B7B Microsoft_VisualStudio_LanguageServices!Microsoft.VisualStudio.LanguageServices.Implementation.Progression.AbstractGraphProvider.BeginGetGraphData(Microsoft.VisualStudio.GraphModel.IGraphContext)+0x3b
    16CCE63C 14569AC2 EditorTestApp!Microsoft.VisualStudio.Progression.ContextWrapper.BeginGetGraphData()+0x52
~~~

**Comment by balajikris** [comment]
On 2/25/, http://dotnet-ci.cloudapp.net/job/roslyn-internal_stabil_win_dbg_eta/151/consoleFull#-202045483561aefe2b-268e-4e71-8c03-11b4c0acbfbc
~~~
00:16:12 System.AggregateException: Exceptions found in the exception log ---> System.InvalidOperationException: The calling thread cannot access this object because a different thread owns it.
00:16:12    at System.Windows.Threading.Dispatcher.VerifyAccess()
00:16:12    at System.Windows.Media.Imaging.BitmapSource.WeakBitmapSourceEventSink.AttachSourceDownloadHandlers()
00:16:12    at System.Windows.Media.Imaging.BitmapSource.WeakBitmapSourceEventSink.set_EventSource(BitmapSource value)
00:16:12    at System.Windows.Media.Imaging.BitmapSource.RegisterDownloadEventSource(BitmapSource eventSource)
00:16:12    at System.Windows.Media.Imaging.CroppedBitmap.SourcePropertyChangedHook(DependencyPropertyChangedEventArgs e)
00:16:12    at System.Windows.Media.Imaging.CroppedBitmap.SourcePropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
00:16:12    at System.Windows.DependencyObject.OnPropertyChanged(DependencyPropertyChangedEventArgs e)
00:16:12    at System.Windows.Freezable.OnPropertyChanged(DependencyPropertyChangedEventArgs e)
00:16:12    at System.Windows.DependencyObject.NotifyPropertyChange(DependencyPropertyChangedEventArgs args)
00:16:12    at System.Windows.DependencyObject.UpdateEffectiveValue(EntryIndex entryIndex, DependencyProperty dp, PropertyMetadata metadata, EffectiveValueEntry oldEntry, EffectiveValueEntry& newEntry, Boolean coerceWithDeferredReference, Boolean coerceWithCurrentValue, OperationType operationType)
00:16:12    at System.Windows.DependencyObject.SetValueCommon(DependencyProperty dp, Object value, PropertyMetadata metadata, Boolean coerceWithDeferredReference, Boolean coerceWithCurrentValue, OperationType operationType, Boolean isInternal)
00:16:12    at System.Windows.DependencyObject.SetValueInternal(DependencyProperty dp, Object value)
00:16:12    at System.Windows.Media.Imaging.CroppedBitmap..ctor(BitmapSource source, Int32Rect sourceRect)
00:16:12    at Microsoft.VisualStudio.Language.Intellisense.Implementation.StandardGlyphProvider.GetGlyph(StandardGlyphGroup group, StandardGlyphItem item)
00:16:12    at Microsoft.VisualStudio.LanguageServices.Implementation.Progression.IconHelper.<>c__DisplayClass2_1.<Initialize>b__0() in d:\j\workspace\roslyn-intern---7f5f7a72\Open\src\VisualStudio\Core\Def\Implementation\Progression\IconHelper.cs:line 83
00:16:12    at Roslyn.Editor.TestApp.UI.ToolWindows.GraphNodeItemHost.Microsoft.VisualStudio.Progression.IIconService.AddIcon(String imageName, String label, Func`1 imageCreator) in d:\j\workspace\roslyn-intern---7f5f7a72\Closed\Hosting\EditorTestApp\UI\ToolWindows\SolutionExplorer\GraphProvider\Host\GraphNodeItemHost.cs:line 118
00:16:12    at Microsoft.VisualStudio.LanguageServices.Implementation.Progression.IconHelper.Initialize(IGlyphService glyphService, IIconService iconService) in d:\j\workspace\roslyn-intern---7f5f7a72\Open\src\VisualStudio\Core\Def\Implementation\Progression\IconHelper.cs:line 83
00:16:12    at Microsoft.VisualStudio.LanguageServices.Implementation.Progression.AbstractGraphProvider.EnsureInitialized() in d:\j\workspace\roslyn-intern---7f5f7a72\Open\src\VisualStudio\Core\Def\Implementation\Progression\GraphProvider.cs:line 46
00:16:12    at Microsoft.VisualStudio.LanguageServices.Implementation.Progression.AbstractGraphProvider.BeginGetGraphData(IGraphContext context) in d:\j\workspace\roslyn-intern---7f5f7a72\Open\src\VisualStudio\Core\Def\Implementation\Progression\GraphProvider.cs:line 136
00:16:12    at Microsoft.VisualStudio.Progression.ContextWrapper.BeginGetGraphData() in d:\j\workspace\roslyn-intern---7f5f7a72\Closed\Hosting\EditorTestApp\Graph\ContextWrapper.cs:line 142
00:16:12    --- End of inner exception stack trace ---
00:16:12    at Roslyn.Hosting.Test.CloseTargetAction.ExecuteOnUIThread() in d:\j\workspace\roslyn-intern---7f5f7a72\Closed\Hosting\Test\Actions\EditorTestApp\CloseTargetAction.cs:line 16
00:16:12    at Roslyn.Hosting.Test.EditorTestAppAction.<>c__DisplayClass25_0.<ExecuteOnUIThread>b__0() in d:\j\workspace\roslyn-intern---7f5f7a72\Closed\Hosting\Test\Actions\EditorTestApp\EditorTestAppAction.cs:line 165
00:16:12 --- End of stack trace from previous location where exception was thrown ---
00:16:12    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
00:16:12    at Roslyn.Hosting.Test.EditorTestAppAction.ExecuteOnUIThread(Action a) in d:\j\workspace\roslyn-intern---7f5f7a72\Closed\Hosting\Test\Actions\EditorTestApp\EditorTestAppAction.cs:line 178
00:16:12    at Roslyn.Hosting.Test.EditorTestAppAction.Execute(TaoEngine engine) in d:\j\workspace\roslyn-intern---7f5f7a72\Closed\Hosting\Test\Actions\EditorTestApp\EditorTestAppAction.cs:line 147
00:16:12    at Tao.Engine.DebuggableActionExecutor.Execute() in f:\dd\src\Omni\Tao\Tao.Engine\ActionExecutor.cs:line 139
00:16:12    at Tao.Engine.DefaultActionExecutor.Execute() in f:\dd\src\Omni\Tao\Tao.Engine\ActionExecutor.cs:line 36
00:16:12 ---> (Inner Exception #0) System.InvalidOperationException: The calling thread cannot access this object because a different thread owns it.
00:16:12    at System.Windows.Threading.Dispatcher.VerifyAccess()
00:16:12    at System.Windows.Media.Imaging.BitmapSource.WeakBitmapSourceEventSink.AttachSourceDownloadHandlers()
00:16:12    at System.Windows.Media.Imaging.BitmapSource.WeakBitmapSourceEventSink.set_EventSource(BitmapSource value)
00:16:12    at System.Windows.Media.Imaging.BitmapSource.RegisterDownloadEventSource(BitmapSource eventSource)
00:16:12    at System.Windows.Media.Imaging.CroppedBitmap.SourcePropertyChangedHook(DependencyPropertyChangedEventArgs e)
00:16:12    at System.Windows.Media.Imaging.CroppedBitmap.SourcePropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
00:16:12    at System.Windows.DependencyObject.OnPropertyChanged(DependencyPropertyChangedEventArgs e)
00:16:12    at System.Windows.Freezable.OnPropertyChanged(DependencyPropertyChangedEventArgs e)
00:16:12    at System.Windows.DependencyObject.NotifyPropertyChange(DependencyPropertyChangedEventArgs args)
00:16:12    at System.Windows.DependencyObject.UpdateEffectiveValue(EntryIndex entryIndex, DependencyProperty dp, PropertyMetadata metadata, EffectiveValueEntry oldEntry, EffectiveValueEntry& newEntry, Boolean coerceWithDeferredReference, Boolean coerceWithCurrentValue, OperationType operationType)
00:16:12    at System.Windows.DependencyObject.SetValueCommon(DependencyProperty dp, Object value, PropertyMetadata metadata, Boolean coerceWithDeferredReference, Boolean coerceWithCurrentValue, OperationType operationType, Boolean isInternal)
00:16:12    at System.Windows.DependencyObject.SetValueInternal(DependencyProperty dp, Object value)
00:16:12    at System.Windows.Media.Imaging.CroppedBitmap..ctor(BitmapSource source, Int32Rect sourceRect)
00:16:12    at Microsoft.VisualStudio.Language.Intellisense.Implementation.StandardGlyphProvider.GetGlyph(StandardGlyphGroup group, StandardGlyphItem item)
00:16:12    at Microsoft.VisualStudio.LanguageServices.Implementation.Progression.IconHelper.<>c__DisplayClass2_1.<Initialize>b__0() in d:\j\workspace\roslyn-intern---7f5f7a72\Open\src\VisualStudio\Core\Def\Implementation\Progression\IconHelper.cs:line 83
00:16:12    at Roslyn.Editor.TestApp.UI.ToolWindows.GraphNodeItemHost.Microsoft.VisualStudio.Progression.IIconService.AddIcon(String imageName, String label, Func`1 imageCreator) in d:\j\workspace\roslyn-intern---7f5f7a72\Closed\Hosting\EditorTestApp\UI\ToolWindows\SolutionExplorer\GraphProvider\Host\GraphNodeItemHost.cs:line 118
00:16:12    at Microsoft.VisualStudio.LanguageServices.Implementation.Progression.IconHelper.Initialize(IGlyphService glyphService, IIconService iconService) in d:\j\workspace\roslyn-intern---7f5f7a72\Open\src\VisualStudio\Core\Def\Implementation\Progression\IconHelper.cs:line 83
00:16:12    at Microsoft.VisualStudio.LanguageServices.Implementation.Progression.AbstractGraphProvider.EnsureInitialized() in d:\j\workspace\roslyn-intern---7f5f7a72\Open\src\VisualStudio\Core\Def\Implementation\Progression\GraphProvider.cs:line 46
00:16:12    at Microsoft.VisualStudio.LanguageServices.Implementation.Progression.AbstractGraphProvider.BeginGetGraphData(IGraphContext context) in d:\j\workspace\roslyn-intern---7f5f7a72\Open\src\VisualStudio\Core\Def\Implementation\Progression\GraphProvider.cs:line 136
00:16:12    at Microsoft.VisualStudio.Progression.ContextWrapper.BeginGetGraphData() in d:\j\workspace\roslyn-intern---7f5f7a72\Closed\Hosting\EditorTestApp\Graph\ContextWrapper.cs:line 142<---
~~~

**Comment by balajikris** [comment]
Found the issue:

ETA's implementation of MS.VS.Progression.IconService invokes the func and creates the image whereas VS' implementation does not realize the func immediately, its lazy. Thus ETA ends up creating UIElements on background threadpool threads and gets lucky or unlucky when it tries to access them again. 

```C#
ETA:
        void IIconService.AddIcon(string imageName, string label, Func<ImageSource> imageCreator)
        {
            icons[imageName] = imageCreator();
        }

VS:
        public override void AddIcon(string imageName, string label, Func<ImageSource> imageCreator)
        {
            IVsUIObject uiObj = new LazyImageObject(imageCreator);
```

**Comment by balajikris** [comment]
Fixed via dotnet/roslyn-internal#682
